Support deferred blocks in partial compilation

crisbeto opened the pull request:
Includes some changes that are necessary to support `@defer` blocks in partial compilation mode.

### refactor(compiler): ensure defer blocks maintain the template order in the binder

Switches to tracking the deferred blocks in a flat array in the binder to ensure that their template order is maintained. This will be relevant in the next commits where we'll match deferred blocks by their index.

Note that the current map we have technically guarantees the insertion order in the spec, but the array is a bit more explicit.


### refactor(compiler): widen type of resolver function
Updates the type of the resolver function to be any `Expression` since JIT may receive a function reference rather than a `ArrowFunctionExpr`.


### refactor(compiler-cli): add method for extracting function parameters
Updates the linker's AST host to add the ability to extract a function's parameters. This will be required for the next commits.


### refactor(core): introduce API for declaring asynchronous partial metadata
Adds the `ɵɵngDeclareClassMetadataAsync` function that will be produced during partial compilation for component classes that have deferred dependencies. At runtime the dependencies will be resolved before setting the metadata.


### feat(compiler-cli): add partial compilation support for deferred blocks
Builds on top of the previous changes to add support for deferred blocks during partial compilation. To do this, the following changes had to be made:
* The metadata passed into `ɵɵngDeclareComponent` has an additional field called `deferBlockDependencies` which has an array of the dependency loading functions for each defer block in the template. During linking, the dependency functions are loaded by matching their template index to the index in the `deferBlockDependencies` array.
* There's a new `ɵɵngDeclareClassMetadataAsync` function that is created for components that have deferred dependencies. It gets transpiled to `setClassMetadataAsync` and works in the same way by capturing a dependency loading function and setting the metadata after the dependencies are resolved. It also has some extra fields for capturing the version which are standard in linker-generated code.
* Deferred import statements are now stripped in partial compilation mode, similar to full compilation.

crisbeto committed:
refactor(compiler): ensure defer blocks maintain the template order in the binder

Switches to tracking the deferred blocks in a flat array in the binder to ensure that their template order is maintained. This will be relevant in the next commits where we'll match deferred blocks by their index.

Note that the current map we have technically guarantees the insertion order in the spec, but the array is a bit more explicit.

crisbeto committed:
refactor(compiler): widen type of resolver function

Updates the type of the resolver function to be any `Expression` since JIT may receive a function reference rather than a `ArrowFunctionExpr`.

crisbeto committed:
refactor(compiler-cli): add method for extracting function parameters

Updates the linker's AST host to add the ability to extract a function's parameters. This will be required for the next commits.

crisbeto committed:
refactor(core): introduce API for declaring asynchronous partial metadata

Adds the `ɵɵngDeclareClassMetadataAsync` function that will be produced during partial compilation for component classes that have deferred dependencies. At runtime the dependencies will be resolved before setting the metadata.

AndrewKushnir reviewed (approved):
@crisbeto the change looks great 👍 Huge thanks for splitting the change into multiple commits, it made it much easier to review.

AndrewKushnir commented:
@alxhub could you please take a look at this change when you get a chance? Would love to get your feedback on **the shape of partially compiled components**. Thank you.

JoostK commented on packages/compiler/src/render3/partial/component.ts:
```diff
@@ -117,6 +117,29 @@ export function createComponentDefinitionMap(
+    let hasResolvers = false;
+
+    for (const deps of meta.defer.blocks.values()) {
+      // Note: we need to push a `null` even if there are no dependencies, because matching of
+      // defer resolver functions to defer blocks happens by index and not adding an array
+      // entry for a block can throw off the blocks coming after it.
+      if (deps === null) {
+        resolvers.push(o.literal(null));
```
This looks sound, but I couldn't find such occurrence in the golden files.

crisbeto replied:
I've added a test that captures it.

JoostK reviewed (approved):
LGTM with one comment

crisbeto committed:
feat(compiler-cli): add partial compilation support for deferred blocks

Builds on top of the previous changes to add support for deferred blocks during partial compilation. To do this, the following changes had to be made:
* The metadata passed into `ɵɵngDeclareComponent` has an additional field called `deferBlockDependencies` which has an array of the dependency loading functions for each defer block in the template. During linking, the dependency functions are loaded by matching their template index to the index in the `deferBlockDependencies` array.
* There's a new `ɵɵngDeclareClassMetadataAsync` function that is created for components that have deferred dependencies. It gets transpiled to `setClassMetadataAsync` and works in the same way by capturing a dependency loading function and setting the metadata after the dependencies are resolved. It also has some extra fields for capturing the version which are standard in linker-generated code.
* Deferred import statements are now stripped in partial compilation mode, similar to full compilation.

dylhunn commented:
This PR was merged into the repository by commit 5bd188a394d30053099e2c83fe79136d590e5399.
